src/qibo/backends/numpy.py

```diff
@@ -769,12 +769,31 @@ def calculate_matrix_exp(self, a, matrix, eigenvectors=None, eigenvalues=None):
         ud = self.np.transpose(np.conj(eigenvectors))
         return self.np.matmul(eigenvectors, self.np.matmul(expd, ud))
 
-    def calculate_matrix_power(self, matrix, power: Union[float, int]):
+    def calculate_matrix_power(
+        self,
+        matrix,
+        power: Union[float, int],
+        precision_singularity: float = 1e-14,
+    ):
         if not isinstance(power, (float, int)):
             raise_error(
                 TypeError,
                 f"``power`` must be either float or int, but it is type {type(power)}.",
             )
+
+        if power < 0.0:
+            # negative powers of singular matrices via SVD
+            # det with np instead of self.np because of torch
+            determinant = np.linalg.det(matrix)
```

Comment:
is this going to work with `CupyBackend`? 
But in general why don't we use the usual `self.np`?

Reply:
It is working locally

Reply:
> In this specific point of the code the matrix might be a torch.Tensor or np.ndarray

but why is that the case? I mean the `backend.calculate_matrix_power` method should expect native arrays of the backend as inputs, right?